Draft RPC Doc

xf6wang opened the pull request:
WIP. previous pr: https://github.com/xf6wang/ts-docs/pull/1

xf6wang committed:
rpc doc

xf6wang commented:
<img width="1464" alt="screen shot 2018-02-28 at 4 58 00 pm" src="https://user-images.githubusercontent.com/22967511/36821624-b23f650e-1ca8-11e8-9525-ade11bbc3049.png">

need to center some things

xf6wang commented:
![image_rpc_state](https://user-images.githubusercontent.com/22967511/36821698-24d51488-1ca9-11e8-8aa4-7b48ffa504ba.png)

xf6wang commented:
looks different between svg and png...

xf6wang commented:
<img width="1717" alt="screen shot 2018-03-01 at 11 22 26 am" src="https://user-images.githubusercontent.com/22967511/36864903-e46af288-1d42-11e8-89ef-3a390bb21316.png">

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,363 @@
+Overview
+========
+
+To communicate between ``traffic_manager`` and ``traffic_server`` there is an RPC mechanism in
```
Use
```
program:`traffic_manager`
```
etc.

xf6wang replied:
added.

SolidWallOfCode commented:
Overall comment - this needs to be moved to `doc/developer-guide/internal-libraries` - it's not really part of the documentation of the documentation.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+To communicate between :program:`traffic_manager` and :program:`traffic_server` there is an RPC mechanism in
+``mgmt/``. This is a simple serialization style RPC which runs over sockets.
+
+=======================================================
```
Headlines need to be at least as long as the text. Because you added `:program:` this doesn't fit anymore. I recommend, though, taking that out and having just `Runtime Structure` as the heading or `Remote RPC runtime structure`.

xf6wang replied:
renamed to `Runtime Structure`

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+
+:program:`traffic_manager` opens a socket to recieve commands from remote clients. :program:`traffic_manager` has a socket connection to :program:`traffic_server` to relay RPC commands.
+
+===========================================================
```
Title too long here, see previous comment. "Message Passing" - don't need "Sequence Diagram", that's obvious.

xf6wang replied:
renamed to `Message Passing`

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+    * MGMT_MARSHALL_STRING: Check string is NULL terminated. Read 4 bytes to determine size of string. Copy entire string. 
+    * MGMT_MARSHALL_DATA: Read 4 bytes to determine data size in bytes. Copy all bytes of data.  
+
+========================================================================
```
Title too long.

xf6wang replied:
changed to `Local Serialization`

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+Overview
+========
+
+To communicate between :program:`traffic_manager` and :program:`traffic_server` there is an RPC mechanism in
```
This should be more broad, given the diagram that's immediately below. Talk about the entire message passing structure in concert with that diagram so they are coherent.

xf6wang replied:
updated.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+   [traffic_cop] <-r-> [traffic_manager] : Remote RPC
+   [other remote clients] <-u-> [traffic_manager] : Remote RPC
+   [traffic_manager] <-r-> [traffic_server] : Local RPC
+   [traffic_server] <-r-> [plugin]
```
Label this `Hook`

xf6wang replied:
done.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+   [traffic_manager] <-r-> [traffic_server] : Local RPC
+   [traffic_server] <-r-> [plugin]
+
+:program:`traffic_manager` opens a socket to recieve commands from remote clients. :program:`traffic_manager` has a socket connection to :program:`traffic_server` to relay RPC commands.
```
What type of socket is it? TCP, Unix domain, pipe?

xf6wang replied:
updated.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+
+Sequence diagram for a command sent from :program:`traffic_ctl` to when it is recieved by a plugin. 
+
+.. image:: ../../uml/images/RPC-sequence-diagram.svg
```
Need to get this image bigger somehow. It's not quite legible.

xf6wang replied:
the sizing seems limited by margins of the html page. We can split this up if need be.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+
+The RPC API for remote clients, such as :program:`traffic_ctl`, :program:`traffic_cop`, etc, is different from the RPC API used between :program:`traffic_manager` and :program:`traffic_server`. 
+
+:program:`traffic_manager` acts like a bridge for remote clients to interact with :program:`traffic_server`. Thus, it is currently impossible to do things like have :program:`traffic_ctl` directly send messages to :program:`traffic_server`. Classes suffixed with "Remote", ie. :file:`CoreAPIRemote.cc`, and classes suffixed with "Local", ie. :file:`NetworkUtilsLocal.cc` are for remote and local cli ...
```
For source files, you should use `:ts:git:` instead of `:file:`. E.g.
```
:ts:git:`mgmt/api/CoreAPIRemote.cc`
```

xf6wang replied:
done.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+Remote RPC vs Local RPC 
+=======================
+
+The RPC API for remote clients, such as :program:`traffic_ctl`, :program:`traffic_cop`, etc, is different from the RPC API used between :program:`traffic_manager` and :program:`traffic_server`. 
```
You might look at local substitutions for these programs, as is done in `common.defs` but just in this file locally.

xf6wang replied:
updated.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+
+   This is the class used to marshall data objects. It provides functions to marshall and unmarshall data. Each data object is associated with a field. Fields are ``MgmtMarshallType``:
+
+.. type:: MgmtMarshallType
```
This should all be indented to indicate it's part fo `MgmtMarshall`. Note that for documentation we use 3 spaces for indent (because almost all the time it's after `.. `).

xf6wang replied:
indented.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+.. c:macro:: MGMT_MARSHALL_STRING
+.. c:macro:: MGMT_MARSHALL_DATA
+
+Serialized data is sent over sockets. This is the general serialization mechanism for RPC communcation. 
```
"When data is actually sent over a connection it must first be serialized".

xf6wang replied:
updated.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+Marshalling:
+============
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_marshall(void *buf, size_t remain, const MgmtMarshallType *fields, unsigned count, ...)
```
Non doc note - this would be so much better as a variadic template.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+
+   This function goes through all the data objects and serializes them together into a buffer. Based on the field, the number of bytes is determined and if there is enough space, it is written into the buffer. 
+
+    * MGMT_MARSHALL_INT: 4 bytes.
```
These should have the same indent level as the preceeding paragraph.

xf6wang replied:
updated.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+   
+   This function parses all the serialized. Based on the field, the number of bytes to be read is determined and copied into a ``MgmtMarshallAnyPtr``.
+
+    * MGMT_MARSHALL_INT: 4 bytes.
```
Are these different from the ones in the previous marshaling section? If not, it should be broken out elsewhere in a more general discussion of serialized data formatting.

xf6wang replied:
It specifies the composition of each `MgmtMarshallType`. I had it for marshalling and unmarshalling but makes more sense to just describe the macro.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+
+State Machine diagrams for :program:`traffic_server` and :program:`traffic_manager` RPC communication. Messages between :program:`traffic_manager` and :program:`traffic_server` begin with a ``MgmtMessageHdr`` followed by the data. 
+
+.. image:: ../../uml/images/RPC-states.svg
```
Drop the double nesting of the outer most states - combine the titles if you think that's required. Or add a caption.

xf6wang replied:
dropped.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+
+.. class:: MgmtMessageHdr
+
+.. member:: int msg_id 
```
This needs to be indented to be part of `MgmtMessageHdr`.

xf6wang replied:
done.

SolidWallOfCode commented on doc/developer-guide/documentation/RPC.en.rst:
```diff
@@ -0,0 +1,397 @@
+
+   #. ``read(int fd, void *buf, size_t count)`` <- a C system call.
+
+2. Read the content:
```
`#.` not `2.`.

xf6wang replied:
done.

SolidWallOfCode commented:
This is just a first pass. I'll have more to say later.

xf6wang commented:
updated it to address the comments.

xf6wang commented on doc/developer-guide/internal-libraries/RPC.en.rst:
```diff
@@ -0,0 +1,403 @@
+
+   This class implements a mgmt_message_sender and specifies a function to send messages to |TManager|.
+
+.. function:: TSMgmtError mgmtapi_sender::send(void *msg, size_t msglen) const override;
```
should I also indent these functions?

SolidWallOfCode replied:
Indenting is a scoping indicator. Functions in classes are indented to show they are part of the class. Global functions in general will not be indented. In this case, the `function` should be indented and inside a `class:: mgmtapi_sender` (unless that's a namespace).

xf6wang commented on doc/developer-guide/internal-libraries/RPC.en.rst:
```diff
@@ -0,0 +1,403 @@
+
+.. _developer-doc-rpc:
+
+.. |TServer| replace:: :program:`traffic_server`
```
I see there's a `|TS|` replace. Should I be using this instead? If so, does `traffic_manager` need to look like Traffic Manager for consistency.

SolidWallOfCode replied:
No, because `|TS|` is for the generic project name, but these are specific executables and should be marked as such.

xf6wang commented:
<img width="1571" alt="screen shot 2018-03-12 at 12 12 31 pm" src="https://user-images.githubusercontent.com/22967511/37304419-c37f0d52-25ee-11e8-97ba-4bdf48a64ea8.png">

SolidWallOfCode commented:
Also, you'll need to add type definitions to `doc/developer-guide/api/types/SystemTypes.en.rst` to remove the doc generation warnings like
```
/home/amc/git/pr/doc/developer-guide/internal-libraries/RPC.en.rst:118: WARNING: cpp:typeOrConcept reference target not found: ssize_t
```
I think once that's done we can put it to ASF.

SolidWallOfCode commented:
You also need to update the `index.en.rst` file in the directory to include `RPC.en`.

SolidWallOfCode commented on doc/developer-guide/internal-libraries/RPC.en.rst:
```diff
@@ -0,0 +1,403 @@
+   a. data length is given in the header, 
+
+   #. read data_len bytes with the same proceedure as the MgmtMessageHdr.
+
```
Instead of this list, you might try something like
```
.. uml::

   partition "Read MgmtMessageHdr" {
   :mgmt_read_pipe();
   -> calls;
   :read_socket();
   -> calls;
   :read();
   note right: System call
   }

   partition "Read message content" {
   :Get content size from header;
   :mgmt_read_pipe();
   -> calls;
   :read_socket();
   -> calls;
   :read();
   note right: System call
   }
```
This is "Activity Diagram" (beta) in PlantUML.

xf6wang commented:
@SolidWallOfCode did a few things and squashed everything. If you could give it one more scan and if it looks good, I'll put it up on open-source. 

1. Added `TSMgmtTypes.en.rst` in /api/types/ and moved macros there
2. Moved rpc.en.rst into a new folder `core-architecture` (I think you mentioned this in the previous doc PR I had)
3. Updated it to reflect the PR that removes the delay from `traffic_ctl`.

SolidWallOfCode commented:
Looks good enough for public review, close this and put it on the ASF github.
